In [ ]:
# importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder 
le = LabelEncoder()

# importing the dataset
df  = pd.read_csv("Iris.csv")
df.head()

In [ ]:
df = df.drop(["Id"],axis=1)
df.head(100)

In [118]:
df["Species"] = le.fit_transform(df["Species"])
df.head()

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
0            5.1           3.5            1.4           0.2        0
1            4.9           3.0            1.4           0.2        0
2            4.7           3.2            1.3           0.2        0
3            4.6           3.1            1.5           0.2        0
4            5.0           3.6            1.4           0.2        0

In [119]:

df.info()
df.head(75)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SepalLengthCm  150 non-null    float64
 1   SepalWidthCm   150 non-null    float64
 2   PetalLengthCm  150 non-null    float64
 3   PetalWidthCm   150 non-null    float64
 4   Species        150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
0             5.1           3.5            1.4           0.2        0
1             4.9           3.0            1.4           0.2        0
2             4.7           3.2            1.3           0.2        0
3             4.6           3.1            1.5           0.2        0
4             5.0           3.6            1.4           0.2        0
..            ...           ...            ...           ...      ...
70            5.9           3.2            4.8           1.8        1
71            6.1           2.8            4.0           1.3        1
72            6.3           2.5            4.9           1.5        1
73            6.1           2.8            4.7           1.2        1
74            6.4           2.9            4.3           1.3        1

[75 rows x 5 columns]

Adding the calsifiaction col 

In [120]:
b = []
for i in df["Species"]:
    if i == 1:
        b.append("Iris-versicolor")
    else:
        b.append("Non - Iris-versicolor")
df["claasName"] = b

print(df)
df.head(100)

     SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species  \
0              5.1           3.5            1.4           0.2        0   
1              4.9           3.0            1.4           0.2        0   
2              4.7           3.2            1.3           0.2        0   
3              4.6           3.1            1.5           0.2        0   
4              5.0           3.6            1.4           0.2        0   
..             ...           ...            ...           ...      ...   
145            6.7           3.0            5.2           2.3        2   
146            6.3           2.5            5.0           1.9        2   
147            6.5           3.0            5.2           2.0        2   
148            6.2           3.4            5.4           2.3        2   
149            5.9           3.0            5.1           1.8        2   

                 claasName  
0    Non - Iris-versicolor  
1    Non - Iris-versicolor  
2    Non - Iris-versicol

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species  \
0             5.1           3.5            1.4           0.2        0   
1             4.9           3.0            1.4           0.2        0   
2             4.7           3.2            1.3           0.2        0   
3             4.6           3.1            1.5           0.2        0   
4             5.0           3.6            1.4           0.2        0   
..            ...           ...            ...           ...      ...   
95            5.7           3.0            4.2           1.2        1   
96            5.7           2.9            4.2           1.3        1   
97            6.2           2.9            4.3           1.3        1   
98            5.1           2.5            3.0           1.1        1   
99            5.7           2.8            4.1           1.3        1   

                claasName  
0   Non - Iris-versicolor  
1   Non - Iris-versicolor  
2   Non - Iris-versicolor  
3   Non - Iris-versicolor  
4   Non - Iris-versicolor  
..                    ...  
95        Iris-versicolor  
96        Iris-versicolor  
97        Iris-versicolor  
98        Iris-versicolor  
99        Iris-versicolor  

[100 rows x 6 columns]

changing clasfication to a number 

In [121]:
df["claasName"] = le.fit_transform(df["claasName"])

df = df.drop(["Species"],  axis = 1)

df.info()

df.head(100)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SepalLengthCm  150 non-null    float64
 1   SepalWidthCm   150 non-null    float64
 2   PetalLengthCm  150 non-null    float64
 3   PetalWidthCm   150 non-null    float64
 4   claasName      150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  claasName
0             5.1           3.5            1.4           0.2          1
1             4.9           3.0            1.4           0.2          1
2             4.7           3.2            1.3           0.2          1
3             4.6           3.1            1.5           0.2          1
4             5.0           3.6            1.4           0.2          1
..            ...           ...            ...           ...        ...
95            5.7           3.0            4.2           1.2          0
96            5.7           2.9            4.2           1.3          0
97            6.2           2.9            4.3           1.3          0
98            5.1           2.5            3.0           1.1          0
99            5.7           2.8            4.1           1.3          0

[100 rows x 5 columns]

Spillting the data

In [122]:
X = df.iloc[:,[0,1,2,3]].values
y = df.iloc[:,4].values

X = X.reshape(-1, 4)
X = preprocessing.scale(X) #scale the data so that it is easier to fit

X_train, X_test, y_train, y_test = train_test_split(X, 
                y, test_size=0.25, random_state=0)


In [123]:
# fit a model
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

# make predictions on test data
y_pred = log_reg.predict(X_test).reshape(-1,1)

In [124]:
# Use score method to get accuracy of model
score = log_reg.score(X_test, y_test)

print('Accuracy: {}'.format(score))

Accuracy: 0.6578947368421053


The confusion matrix 

In [125]:
from sklearn.metrics import confusion_matrix

classes = ["Iris-versicolor ","Non - Iris-versicolor"]
conf_mat = confusion_matrix(y_test, y_pred)
cm_df = pd.DataFrame(conf_mat, columns=classes, index=classes)
cm_df

Iris-versicolor   Non - Iris-versicolor
Iris-versicolor                       4                     12
Non - Iris-versicolor                 1                     21

from the confution matrix we can exspect that there would be a high recall value and a slightly above avrage accuracy rating 

In [126]:
from sklearn.metrics import f1_score, precision_score, recall_score

# average f1 score
av_f1 = f1_score(y_test, y_pred, average='micro')
print(av_f1)

# f1 score per class
f = f1_score(y_test, y_pred, average=None)
lowest_score = min(f)
hardest_class = classes[list(f).index(lowest_score)]
print('Hardest class:', hardest_class)

0.6578947368421053
Hardest class: Iris-versicolor 


Score report 

In [127]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.80      0.25      0.38        16
           1       0.64      0.95      0.76        22

    accuracy                           0.66        38
   macro avg       0.72      0.60      0.57        38
weighted avg       0.71      0.66      0.60        38



Here we cans see that that recall value for iris sorta is 0.95 and the pricition is 0.64. 